feat: DateRangePicker Implementation (#780)
deephaven/deephaven-plugins: plugins/ui/docs/components/date_range_picker.md

In addition, when a value with a time is provided but you wish to only display the date, you can set the granularity to "DAY". This has no effect on the actual value (it still has a time component), only on what fields are displayed. In the following example, two date range pickers are synchronized with the same value, but display different granularities.

```python
from deephaven import ui

@ui.component
def granularity_example():
    value, set_value = ui.use_state(
        {"start": "2021-04-07T18:45:22 UTC", "end": "2021-04-08T20:00:00 UTC"}
    )
    return ui.flex(
        ui.date_range_picker(
            label="Date range and time range",
            granularity="SECOND",
            value=value,
            on_change=set_value,
        ),
        ui.date_range_picker(
            label="Date range", granularity="DAY", value=value, on_change=set_value
        ),
        gap="size-150",
        wrap=True,
    )

my_granularity_example = granularity_example()
```

## HTML forms

> Date range pickr supports the start_name and end_name props for integration with HTML forms. The values will be submitted to the server as `ISO 8601` formatted strings according to the granularity of the value. For example, if the date range picker allows selecting only dates then strings such as "2023-02-03" will be submitted, and if it allows selecting times then strings such as "2023-02-03T08:45:00"

Reviewer: ```suggestion
Date range picker supports the `start_name` and `end_name` props for integration with HTML forms. The values will be submitted to the server as `ISO 8601` formatted strings according to the granularity of the value. For example, if the date range picker allows selecting only dates then strings such as "2023-02-03" will be submitted, and if it allows selecting times then strings such as "2023-02-03T08:45:00"
```

Author: Updated

> ```python
> from deephaven import ui
> 
> my_date_range_picker_forms = ui.date_range_picker(
>     label="Trip dates", start_name="startDate", end_name="endDate"
> )
> ```

Reviewer: This example doesn't make a ton of sense in isolation (I know it's copied from Spectrum). I'd wrap this in a `ui.form`, e.g.

````suggestion
```python
from deephaven import ui

my_date_range_picker_forms = ui.form(
    ui.date_range_picker(
        label="Trip dates", start_name="startDate", end_name="endDate"
    ),
    ui.button("Submit", type="submit"),
    on_submit=print
)
```
````

Author: Updated

# Date Range Picker

Date range pickers allow users to select a range dates and times from a pop up calendar.

## Example

```python
from deephaven import ui

my_date_range_picker_basic = ui.date_range_picker(label="Date range")
```

## Date types

> A date range picker can be used to select `range` of dates.

Reviewer 2: ```suggestion
A date range picker can be used to select a `range` of dates.
```

The `range` is a dictionary with a `start` date and an `end` date. e.g. `{ "start": "2024-01-02", "end": "2024-01-05" }`

> The date range picker accepts the following date types as inputs:  

Reviewer 2: The date range picker accepts the following date types as inputs:  
- `None`
- `LocalDate`
- `ZoneDateTime`
- `Instant`
- `int`
- `str`
- `datetime.datetime`
- `numpy.datetime64`
- `pandas.Timestamp`

Author: Fixed

`None`, `LocalDate`, `ZoneDateTime`, `Instant`, `int`, `str`, `datetime.datetime`, `numpy.datetime64`, `pandas.Timestamp`

> The `start` and `end` dates should be input using the same type.

Reviewer 2: ```suggestion
The `start` and `end` dates should be the same type.
```

The input will be converted to one of three Java date types:

> 1. `LocalDate`: A LocalDate is a date without a time zone in the ISO-8601 system, such as "2007-12-03" or "2057-01-28".

Reviewer 2: ```suggestion
1. `LocalDate`: in the ISO-8601 system, a LocalDate is a date without a time zone, such as "2007-12-03" or "2057-01-28".
```

   This will create a date range picker with a granularity of days.
2. `Instant`: An Instant represents an unambiguous specific point on the timeline, such as 2021-04-12T14:13:07 UTC.
   This will create a date range picker with a granularity of seconds in UTC. The time zone will be rendered as the time zone in user settings.
3. `ZonedDateTime`: A ZonedDateTime represents an unambiguous specific point on the timeline with an associated time zone, such as 2021-04-12T14:13:07 America/New_York.
   This will create a date range picker with a granularity of seconds in the specified time zone. The time zone will be rendered as the specified time zone.

The `start` and `end` inputs are coverted according to the following rules:

1. If the input is one of the three Java date types, use that type.

> 2. A date string such as "2007-12-03" will parse to a `LocalDate`

Reviewer 2: ```suggestion
2. A date string such as "2007-12-03" will parse to a `LocalDate`.
```

> 3. A string with a date, time, and timezone such as "2021-04-12T14:13:07 America/New_York" will parse to a `ZonedDateTime`

Reviewer 2: ```suggestion
3. A string with a date, time, and time zone such as "2021-04-12T14:13:07 America/New_York" will parse to a `ZonedDateTime`
```

4. All other types will attempt to convert in this order: `Instant`, `ZonedDateTime`, `LocalDate`

The format of the date range picker and the type of the value passed to the `on_change` handler

> is determined by the type of the following props in order of precedence:

Reviewer 2: ```suggestion
are determined by the type of the following props in order of precedence:
```

Author: handler is singular. I think we should use `is`

Reviewer 2: The format and type of value are determined...?

In most cases, your data will come from and be sent to a server as an `ISO 8601` formatted string.

For `ZonedDateTime` objects, the date range picker will display the specified time zone.

For `Instant` objects, the date range picker will display the time zone from the user settings.

```python
from deephaven import ui
from deephaven.time import to_j_instant

my_zoned_date_time = ui.date_range_picker(
    label="Date range",
    default_value={
        "start": "2022-11-07T00:45 America/Los_Angeles",
        "end": "2022-11-08T11:15 America/Los_Angeles",
    },
)

my_instant = ui.date_range_picker(
    label="Date range",
    default_value={
        "start": to_j_instant("2022-11-07T00:45Z"),
        "end": to_j_instant("2022-11-08T11:15Z"),
    },
)
```

## Granularity

The `granularity` prop allows you to control the smallest unit that is displayed by a date range picker. By default, `LocalDate` values are displayed with "DAY" granularity (year, month, and day), and `ZonedDateTime` and `Instant` values are displayed with "SECOND" granularity.

> In addition, when a value with a time is provided but you wish to only display the date, you can set the granularity to "DAY". This has no effect on the actual value (it still has a time component), only on what fields are displayed. In the following example, two date range pickers are synchronized with the same value, but display different granularities.

Reviewer 2: ```suggestion
In addition, when a value with a time is provided but you wish to display only the date, you can set the granularity to "DAY". This has no effect on the actual value (it still has a time component), only on what fields are displayed. In the following example, two date range pickers are synchronized with the same value but display different granularities.
```

        {"start": "2021-04-07T18:45:22 UTC", "end": "2021-04-08T20:00:00 UTC"}
    )
    return ui.flex(
        ui.date_range_picker(
            label="Date range and time range",
            granularity="SECOND",
            value=value,
            on_change=set_value,
        ),
        ui.date_range_picker(
            label="Date range", granularity="DAY", value=value, on_change=set_value
        ),
        gap="size-150",
        wrap=True,
    )

my_granularity_example = granularity_example()
```

## HTML forms

Date range picker supports the `start_name` and `end_name` props for integration with HTML forms. The values will be submitted to the server as `ISO 8601` formatted strings according to the granularity of the value. For example, if the date range picker allows selecting only dates then strings such as "2023-02-03" will be submitted, and if it allows selecting times then strings such as "2023-02-03T08:45:00"

```python
from deephaven import ui

my_date_range_picker_forms = ui.form(
    ui.date_range_picker(
        label="Trip dates", start_name="startDate", end_name="endDate"
    ),
    ui.button("Submit", type="submit"),
    on_submit=print,
)
```

## Labeling

> A visual label should be provided for the date range picker using the label prop. If the DateRangePicker is required, the `is_required` and `necessity_indicator` props can be used to show a required state.

Reviewer 2: ```suggestion
A visual label should be provided for the date range picker using the `label` prop. If the date range picker is required, the `is_required` and `necessity_indicator` props can be used to show a required state.
```

Using the `contextual_help` prop, a `ui.contextual_help` can be placed next to the label to provide additional information about the date range picker.

```python
from deephaven import ui

date_range_picker_contextual_help_example = ui.date_range_picker(
    label="Sample Label",
    contextual_help=ui.contextual_help(ui.heading("Content tips")),
)
```

## Custom width

The `width` prop adjusts the width of a date range picker, and the `max_width` prop enforces a maximum width.

```python
from deephaven import ui

@ui.component
def date_range_picker_width_examples():
    return [
        ui.date_range_picker(
            width="size-6000",
        ),
        ui.date_range_picker(
            width="size-6000",
            max_width="100%",
        ),
    ]

my_date_range_picker_width_examples = date_range_picker_width_examples()
```

## Maximum visible months

By default, the calendar popover displays a single month. The max_visible_months prop allows displaying up to 3 months at a time, if screen space permits.

```python
from deephaven import ui

date_range_picker_months_example = ui.date_range_picker(
    label="Date range", max_visible_months=3
)
```

## Page behavior

> By default, when pressing the next or previous buttons, pagination will advance by the max_visible_months value. This behavior can be changed to page by single months instead, by setting page_behavior to single.

Reviewer 2: ```suggestion
By default, when pressing the next or previous buttons, pagination advances by the `max_visible_months` value. This behavior can be changed to page by single months by setting `page_behavior` to single.
```

```python
from deephaven import ui

date_range_picker_page_example = ui.date_range_picker(
    label="Date range", max_visible_months=3, page_behavior="single"
)
```

## Hide time zone

> The time zone can be hidden using the hide_time_zone option.

Reviewer 2: ```suggestion
The time zone can be hidden using the `hide_time_zone` option.
```

```python
from deephaven import ui

my_hide_time_zone_example = ui.date_range_picker(
    label="Date range",
    default_value={
        "start": "2022-11-07T00:45 America/Los_Angeles",
        "end": "2022-11-08T11:15 America/Los_Angeles",
    },
    hide_time_zone=True,
)
```

## Hour cycle

> By default, date range picker displays times in either `12` or `24` hour hour format depending on the user's locale. However, this can be overridden using the hour_cycle prop.

Reviewer 2: ```suggestion
By default, date range picker displays times in either a `12` or `24` hour format depending on the user's locale. However, this can be overridden using the `hour_cycle` prop.
```

```python
from deephaven import ui

date_range_picker_hour_cycle_example = ui.date_range_picker(
    label="Date range", hour_cycle=24
)
```

## Time table filtering

> Date Range Pickers can be used to filter tables with time columns.

Reviewer 2: ```suggestion
Date range pickers can be used to filter tables with time columns.
```
